src/lib/agent.c

```diff
@@ -66,7 +66,7 @@ static enum task_state validate(void *data)
 	/* warning timeout */
 	if (delta > sa->warn_timeout)
 		tr_warn(&sa_tr, "validate(), ll drift detected, delta = %u",
-			delta);
+			(int)delta);
```

Comment:
I really don't get why we want to trace 64-bit variables with 32 bits in the first place.
Beyond the GCC issue, at some point this will break anyways.

Reply:
@plbossart I don't really. `%u` is `unsigned integer`, so, it's 32 bits. So the current print is already printing 32 bits (at most), but due to what seems like an xtensa gcc bug it doesn't work even in a "normal" automatic type-cast sense. I would be equally happy to remove this `tr_warn()`, but if we have it, better have at least somewhat useful. And in fact at least in my tests those values were indeed well within 32 bits, so, prints with this change were quite appropriate.

Reply:
@lyakh It isn't a bug specifically. It's just that there is nothing to convert the 64-bit integer to a 32-bit integer before pushing as a vararg. You push a 64-bit integer into varargs and extract 32-bit. Smaller integers do get promoted to 32-bit when considering varargs so that's why you _think_ there is an automatic conversion. Well, everything (even char) is converted to int, and %c just does the opposite conversion.

Reply:
@paulstelian97 I assumed that those who implemented this also used those trace prints and they were producing correct values. Then, seeing, that it doesn't work for me, I considered what could've been the difference. I thought that a likely difference could be, that the original author used xt-xcc. But it does indeed look like I was wrong and probably this actually never worked. Yes, as you say the original code did "push" a 64-bit value to varargs, which isn't a bug per se. `va_arg(3)` says:
> The  va_arg() macro expands to an expression that has the type and value of the next argument in the call.

But then the line
```
	data[PAYLOAD_OFFSET(i)] = va_arg(vl, uint32_t);
```
in `trace_log()` handles that value as a 32-bit one which is wrong:
> if type is not compatible with the type of the actual next argument (as promoted according to the default argument promotions), random errors will occur.

So, looks like we really only should use 32-bit values with tracing, and not even 16- or 8-bit ones without explicit conversions.

Reply:
16-bit and 8-bit integers do work due to automatic conversion to 32-bit when pushing to a vararg (va_arg(..., char) will still extract 4 bytes into an int, then convert said int to char). So they are actually just fine. Just use the correct specifiers (such as 0x%04x or something) and you're good to go.

Reply:
my point was: do we actually want to trace the full 64-bit values? If yes, the existing code and the added cast are not helping.

Reply:
If we want full 64-bit we could make a hack of the kind trace(..., "0x%08x%08x", (uint32_t)(value >> 32), (uint32_t)value);

Reply:
@plbossart exactly, do we? :-) @mmaka1 ? I think here a value is useful, and in fact I also would say, that this patch isn't that wrong. The variables should be 64 bit to correctly perform calculations. But the result will "most likely" lie within 32 bits. We could add a check there `if (drift > UINT_MAX) tr_warn("value out of range"); else tr_warn("...");`? And prints in `sa_init()` can just be removed.

Reply:
added checks